File: leetcode-algorithms/040. Combination Sum II/solution.cpp
class Solution {
private:
    void dfs(vector<int>& nums, int target, int index, vector<int>& path, vector<vector<int>> &res){
    if(target<0){
        return;
    }
    if(target==0){
        res.push_back(path);
        return;
    }
    for(int i=index; i<nums.size();i++){
        if(i>index && nums[i]==nums[i-1])
            continue;
        path.push_back(nums[i]);
        dfs(nums, target-nums[i], i+1, path, res);
        path.pop_back();
    }

}
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
    vector<int> path;
    sort(candidates.begin(), candidates.end());
    dfs(candidates, target, 0, path, res);
    return res;
    }
};

File: leetcode-algorithms/040. Combination Sum II/solution.py
def combinationSum(candidates, target):
	"""
	:type candidates: List[int]
	:type target: int
	:rtype: List[List[int]]
	"""
	def dfs(nums,target, index, path, res):
		if target<0:
			return
		if target == 0:
			res.append(path)
			return
		for i in range(index, len(nums)):
			if i > index and nums[i] == nums[i-1]:
				continue
			dfs(nums, target-nums[i], i+1, path + [nums[i]], res)

	res = []
	candidates.sort()
	dfs(candidates, target, 0, [], res)

	return res


if __name__ == '__main__':
	candidates = [2,5,2,1,2]
	target = 5
	res = combinationSum(candidates,target)
	print(res)